x-pack/plugins/fleet/server/services/security/message_signing_service.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { generateKeyPairSync, createSign, randomBytes } from 'crypto';
+
+import type { KibanaRequest } from '@kbn/core-http-server';
+import type { SavedObjectsClientContract } from '@kbn/core-saved-objects-api-server';
+import type { EncryptedSavedObjectsClient } from '@kbn/encrypted-saved-objects-plugin/server';
+import { SECURITY_EXTENSION_ID } from '@kbn/core-saved-objects-server';
+
+import { MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE } from '../../constants';
+import { appContextService } from '../app_context';
+
+interface MessageSigningKeys {
+  private_key: string;
+  public_key: string;
+  passphrase: string;
+}
+
+export interface MessageSigningServiceInterface {
+  getPublicKey(): Promise<string>;
+  sign(message: unknown): Promise<string>;
+  generateKeyPair(providedPassphrase?: string): Promise<void>;
+}
+
+export class MessageSigningService implements MessageSigningServiceInterface {
+  #esoClient: EncryptedSavedObjectsClient;
+  #_soClient: SavedObjectsClientContract | undefined;
+
+  constructor(esoClient: EncryptedSavedObjectsClient) {
+    this.#esoClient = esoClient;
+  }
+
+  public async getPublicKey(): Promise<string> {
+    const { publicKey } = await this.#getCurrentKeyPair();
+    return publicKey;
+  }
+
+  public async sign(message: unknown): Promise<string> {
+    const signer = createSign('SHA256');
+    signer.update(JSON.stringify(message));
```

Comment:
We should change the ```sign``` implementation and pass bytes Buffer that already contains the JSON serialized UTF-8 blob (bytes), because we would need to pass the blob encoded in the payload. I'll elaborate in the longer comment to this POS.